app.coffee

```diff
@@ -73,3 +82,31 @@ server.listen port, host, (error) ->
 
 # Stop huge stack traces in logs from all the socket.io parsing steps.
 Error.stackTraceLimit = 10
+
+
+shutdownCleanly = (signal) ->
+	connectedClients = io.sockets.clients()?.length
+	if connectedClients == 0
+		logger.log("no clients connected, exiting")
+		process.exit()
+	else
+		logger.log {connectedClients}, "clients still connected, not shutting down yet"
+		setTimeout () ->
+			shutdownCleanly(signal)
+		, 10000
+
+forceDrain = ->
+	logger.log {delay_ms:Settings.forceDrainMsDelay}, "starting force drain after timeout"
+	setTimeout ()-> 
+		logger.log "starting drain"
+		DrainManager.startDrain(io, 4)
+	, Settings.forceDrainMsDelay
+
+if Settings.forceDrainMsDelay?
+	Settings.forceDrainMsDelay = parseInt(Settings.forceDrainMsDelay, 10)
+	logger.log forceDrainMsDelay: Settings.forceDrainMsDelay,"forceDrainMsDelay enabled"
+	for signal in ['SIGINT', 'SIGHUP', 'SIGQUIT', 'SIGUSR1', 'SIGUSR2', 'SIGTERM', 'SIGABRT']
```

Comment:
I think we should limit this behaviour to a single signal, and then use that. Based on Nginx convention, `SIGQUIT` is for a 'graceful' shutdown. Or we may have our hands tied by the signal that k8s/docker will send? But we still want a way to terminate the process immediately if we need to I think.

Reply:
(sorry for the unrequested input!)

Reply:
There are multiple signals which k8 will send this container depending on the reason for taking down the container. I know it sends `SIGABRT` for a delete pod request and a `SIGINT` for scale down requests, `SIGHUP` gets sent later on in the delete lifecycle I think. The above is just a catchall I took from doc updater to ensure I don't miss one.

Reply:
ok, cool, thanks, that makes sense 👍